cpp/src/arrow/testing/math.cc

```diff
@@ -0,0 +1,95 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow/testing/math.h"
+
+#include <cmath>
+#include <limits>
+
+#include <gtest/gtest.h>
+
+#include "arrow/util/logging.h"
+
+namespace arrow {
+namespace {
+
+template <typename Float>
+constexpr Float kOneUlp;
+template <>
+constexpr float kOneUlp<float> = 5.9604645e-08f;
+template <>
+constexpr double kOneUlp<double> = 1.1102230246251565e-16;
+
+template <typename Float>
+bool WithinUlpOneWay(Float left, Float right, int n_ulp) {
+  // Ideally these would be compile-time (constexpr) checks
+  DCHECK_LT(Float(1.0) - kOneUlp<Float>, Float(1.0));
+  DCHECK_EQ(std::nextafter(Float(1.0) - kOneUlp<Float>, Float(1.0)), Float(1.0));
+
+  DCHECK_GE(n_ulp, 1);
+
+  if (left == 0) {
+    return left == right;
+  }
+  if (left < 0) {
+    left = -left;
+    right = -right;
+  }
+
+  int left_exp;
+  Float left_mant = std::frexp(left, &left_exp);
+  Float delta = static_cast<Float>(n_ulp) * kOneUlp<Float>;
```

Comment:
Should we compute the `kOneUlp` on the fly instead of pre-defined constants, for the following reasons:
1. The beginning two `DCHECK`s won't be needed;
2. The platform-specific logic will be delegated to `std` (instead of doing runtime checks using `DCHECK`).

Reply:
Well, the checks are really Arrow internal checks so that we catch any potential issues at CI time. They won't be able in production code.

Reply:
So your point IIUC is that, we don't want to fully trust/leverage `std::nextafter` to handle all compiler/hardware floating point issues. So we prefer to have our own `kOneUlp` literal, and conversely, validate a particular compiler + hardware combination works fine (in terms of computing one ulp) at CI time.

Is this the intention?

Reply:
Ah, then my concerns remain. I'm suggesting computing "one ulp" on the fly because:
1. You don't need `DCHECK`s to guard the correctness of the compile-time constants.
2. Better readability - you don't need a tricky name and an elaborating comment for it.
3. Once the `DCHECK` are eliminated, there is only one reference to `kOneUlp` left, so there isn't code duplication (even there is, we can make it a function).
4. The cost of calling `std::nextafter` should be trivial. Even if it has overhead, the function is only used in testing scenarios so performance is less a considration.

What do you think?

Reply:
Hmm, yes, you're probably right. Let's try this.